File: txtscrio/go32/pas/initialize.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure initialize;
begin
  asm
        mov     ah,0fh
        int     10h
        and     al,7fh
        mov     v_mode,al
        mov     DispPg,bh
  end;

  MaxCol := MEM[SEG0040:$4a];
  MaxLn := SUCC(MEM[SEG0040:$84]);
  work_MaxLn  := MaxLn;
  work_MaxCol := MaxCol;
  FillWord(screen_ptr^,MAX_SCREEN_MEM_SIZE DIV 2,$0700);
  dosmemput(v_seg,v_ofs,screen_ptr^,MAX_SCREEN_MEM_SIZE);
end;

File: txtscrio/go32/initialize.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

void initialize (void) {
#if !USE_FPC
  uint16_t orig_fs;
#endif // !USE_FPC

  VBIOS_get_video_mode (&v_mode, &DispPg);

#if !USE_FPC
  orig_fs = _fargetsel ();
  _farsetsel (_dos_ds);
#endif // !USE_FPC

  MaxCol = _farnspeekw (0x44A); // screen width in text columns
  MaxLn  = _farnspeekb (0x484) + 1; // EGA text rows - 1
#if !ADT2PLAY
  work_MaxCol = MaxCol;
  work_MaxLn  = MaxLn;
#endif // !ADT2PLAY

#if !USE_FPC
  _farsetsel (orig_fs);
#endif // !USE_FPC

  memsetw (screen_ptr, 0x0700, MAX_SCREEN_MEM_SIZE / 2);
#if !ADT2PLAY
  dosmemput (screen_ptr, MAX_SCREEN_MEM_SIZE, v_seg * 16 + v_ofs);
#endif // !ADT2PLAY
}
